### ISSUE TITLE
Not Pretty Printing Can Corrupt Generated JSON 

### ISSUE THREAD

**Issue by delvh** [google/gson#1919]
## Usage
```java
new Gson().toJson(<a Map storing all kinds of values and strings as keys>);
```

## GSON version
`2.8.6`, `2.8.7`

## Problem
I have noticed a possibility to corrupt the output of a generated JSON file ~~(presumably if a collection is at the end -  I do not know whether this also applies to `List`s, but it certainly applies to `Set`s)~~.
It seems to occur only when the output of `toJson` is not pretty printed, so far it has not been observed for pretty printed json strings.

In the json file below, you can see that, even though the JSON object has already been closed, something else is still appended, subsequently breaking the JSON file.
~~I've had this problem occur now more than once, and it seems to me as if when there is a collection at the end of the generated json, it always corrupts the file. This is especially true when this collection is not empty, for empty collections it was so far not observed. When any of the primitive values is at the end, it works as expected (including `String`s).~~

```json
{
  "last_file_chooser_dir": "/home/delvh/git/project/target",
  "recent_documents": [
    "/home/delvh/git/project/test1.xml",
    "/home/delvh/git/project/test2.xml",
    "/home/delvh/git/project/test3.xml"
  ],
  "config_version": 1,
  "open_documents": [
    "/home/delvh/git/project/target",
    "/home/delvh/git/project"
  ]
}ect/target","config_version":1}
```
This example has directly been copied from the output of the `toJson` mentioned above, and was thrown into a JSON formatter, so no manual mutation took place.

~~EDIT: After some testing, this problem seems to be fixed in `2.8.7`. At least so far I was unable to reproduce it there.~~
Still persists in `2.8.7`

## Another example
```json
{
  "last_file_chooser_dir": "/home/delvh/git/test/project",
  "recent_documents": [],
  "language": "en",
  "open_paths": [
    "/home/delvh/git/test/project"
  ],
  "open_documents": [],
  "config_version": 1
}project"],"open_documents":["/home/delvh/git/test/project/test.xml"]}
```

**Comment by ogrammer** [comment]
I recognize this problem as it had affected me in the past. I believe that you are using `Files.writeString` or `Files.write`, right? Well, these methods by default only change the beginning of the file meaning if the new JSON is shorter than the JSON already in the file, there will be leftover parts. So to fix this you need to add `StandardOpenOption.TRUNCATE_EXISTING` like for example `Files.writeString(Path.of("abc.json"), myJson, StandardOpenOption.TRUNCATE_EXISTING);`

Hope it helps.

**Comment by delvh** [comment]
Thanks a lot, @ogrammer.
Now I can consistently reproduce this issue, and what you mentioned seems indeed correct.
Who would have known that supplying **less** parameters would solve this issue.
For context, as you mentioned, previously I used
```java
Files.writeString(path, string, StandardOpenOptions.CREATE, StandardOpenOptions.WRITE)
```
But as it seems, correct would have been either
```java
Files.writeString(path, string, StandardOpenOptions.CREATE, StandardOpenOptions.WRITE, StandardOpenOptions.TRUNCATE_EXISITING)
```
or
```java
Files.writeString(path, string)
```
which appear to do the same as write apparently uses exactly these three options if none are given…

In conclusion, this was not the fault of GSON, but of a semi-intuitive approach in `Files.writeString`.

**Comment by ogrammer** [comment]
@delvh I didn't know just using writeString already supplies `CREATE` and `TRUNCATE_EXISTING`, my code will be cleaner now :) We both learned something

**Comment by Shadowrs** [comment]
I sporadically get the same issue:
gson 2.10.1

```
"data": {
    "typeA": [
      {
        "slot": 185,
        "key": "CHARGES",
        "type": "INTEGER",
        "value": 96
      }
    ]
  }
}**_TEGER",
        "value": 96
      }
    ]
  }_**
}
```

However my write code is this:
```
File file = new File("./out.json"));
            try (BufferedWriter writer = new BufferedWriter(new FileWriter(file, false))) {
                writer.write(myGson.toJson(this));
                writer.flush();
            }
```
